Address new changes in captcha specification

pkong-ds opened the pull request:
Spec ref #4282 

## What's in this PR?

- [x] Update captcha configuration
  - [x] ~Support id_allowlist~ To be implemented
  - [x] Support `mode` (only `always` & `never`)
  - [x] Remove `captcha.enabled` from authflow root
- [x] ~Introduce new `risk_assessment` configuration~ To be implemented
   - [x] ~The Go structs~
   - [x] ~The JSON schema~
   - [x] ~Write tests to test for valid and invalid configurations~
- [x] ~Add `risk_assessment` configuration to the root of each flow type configuration~
   - [x] ~Add `risk_assessment` configuration to the branches of identify, create_authenticator, and authenticate~
   - [ ] ~When generating the default flows, take `risk_assessment` configuration into account. Refer to the spec for the details of the behavior~
   - [ ] ~Write tests for the generated flows.~

~Last 2 items pending in https://github.com/authgear/authgear-server/pull/4282/files#r1646534742~

### New Changes as of June 21
- [x] rename `captcha` as `bot_protection`
- [x] Misc config schema updates
- [x] Updates on all test cases

EDIT: Cross out to be implemented items

EDIT 2: Added new section

louischan-oursky commented on pkg/lib/config/authentication_flow_captcha.go:
```diff
@@ -4,12 +4,60 @@ var _ = Schema.Add("AuthenticationFlowCaptcha", `
 	"properties": {
-		"required": { "type": "boolean" }
+		"mode": { "type": "string", "enum": ["never", "always"] },
+		"fail_open": { "type": "boolean" }
```
No need to add this now. Generally, do not add something that is not going to be implemented now

pkong-ds commented on pkg/lib/authenticationflow/declarative/generate_config_account_recovery_flow.go:
```diff
@@ -49,7 +49,10 @@ func GenerateAccountRecoveryFlowConfig(cfg *config.AppConfig) *config.Authentica
-				Mode: config.AuthenticationFlowCaptchaModeNever,
+				Mode: config.AuthenticationFlowCaptchaModeAlways,
+				Provider: &config.AuthenticationFlowCaptchaProvider{
+					Alias: cfg.Captcha.Providers[0].Alias,
```
will use `getCaptchaProviderConfig` over `hasCaptcha`, since it no longer returns a `bool`

pkong-ds commented:
@louischan-oursky updated, thanks!

pkong-ds commented:
This became WIP because spec updated

pkong-ds commented:
@louischan-oursky local test & ci pass, feel free to review :adore:

louischan-oursky commented on pkg/lib/config/authentication_flow_captcha.go:
```diff
@@ -51,7 +51,11 @@ type AuthenticationFlowCaptchaProvider struct {
 type AuthenticationFlowCaptcha struct {
 	Mode     AuthenticationFlowCaptchaMode      `json:"mode,omitempty"`
-	Provider *AuthenticationFlowCaptchaProvider `json:"provider,omitempty"`
+	Provider *AuthenticationFlowCaptchaProvider `json:"provider,omitempty" nullable:"true"`
```
Remove the struct tag `nullable:"true"` here. Instead, we should add `nullable:"true"` to the structs that contain `AuthenticationFlowCaptcha`

louischan-oursky commented on pkg/lib/config/authentication_flow_bot_protection.go:
```diff
@@ -0,0 +1,60 @@
+}
+type AuthenticationFlowBotProtection struct {
+	Mode     AuthenticationFlowBotProtectionMode      `json:"mode,omitempty"`
+	Provider *AuthenticationFlowBotProtectionProvider `json:"provider,omitempty" nullable:"true"`
```
Do not add nullable here. Add it to where `AuthenticationFlowBotProtection` is used in a struct.

louischan-oursky commented on pkg/lib/config/authentication_flow_bot_protection.go:
```diff
@@ -0,0 +1,60 @@
+
+func (c *AuthenticationFlowBotProtection) SetDefaults() {
+	c.Mode = AuthenticationFlowBotProtectionModeNever
+}
```
Remove SetDefaults() on this struct.

pkong-ds commented:
updated, thanks!

Added 2 commits, starting from [9c41c18](https://github.com/authgear/authgear-server/pull/4382/commits/9c41c18afb38742adc1f950a598bc708c3cf329c)

louischan-oursky commented on pkg/lib/config/bot_protection.go:
```diff
@@ -40,7 +40,7 @@ var _ = Schema.Add("BotProtectionProvider", `
 type BotProtectionConfig struct {
 	Enabled  *bool                  `json:"enabled,omitempty"`
-	Provider *BotProtectionProvider `json:"provider,omitempty" nullable:"true"`
+	Provider *BotProtectionProvider `json:"provider,omitempty"`
```
Provider is nullable. And Please implement SetDefaults() on BotProtectionConfig to set Enabled to false if it is nil

pkong-ds replied:
As per offline discussion, will make Enabled as non-pointer instead

pkong-ds commented:
@louischan-oursky updated, thanks! local ci & e2e passed :adore:
